### DISCUSSION TITLE
response.set_cookie() does not work

### DISCUSSION THREAD

**Discussion by Naxii-e** [fastapi/fastapi#6382] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [x] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Response

app = FastAPI()

@app.get('/cookie')
async def cookie(response: Response):

 # Not work
 response.set_cookie(key="test", value="example")
 return {"result": "bad"}

----------------------------------

from fastapi import FastAPI
from fastapi.responses import Response

app = FastAPI()

@app.get('/cookie')
async def cookie(response: Response):

 # It work
 response.set_cookie(key="test", value="example")
 return {"result": "ok"}
```

### Description

I am new to Python and FastAPI.
I would like to use `response.set_cookie()`.

I wrote the code as per [this](https://fastapi.tiangolo.com/ja/advanced/response-cookies/#response-cookies) document.

```python
from fastapi import FastAPI, Response

app = FastAPI()

@app.get('/cookie')
async def cookie(response: Response):

 # Not work
 response.set_cookie(key="test", value="example")
 return {"result": "bad"}
```
But this did not work. The cookie is not set.

but changed the `Response` import to `from fastapi.responses import Response`.

```python
from fastapi import FastAPI
from fastapi.responses import Response

app = FastAPI()

@app.get('/cookie')
async def cookie(response: Response):

 # It work
 response.set_cookie(key="test", value="example")
 return {"result": "ok"}
```
What are these differences?
Are they bugs or specifications?

Sorry for the confusing text.

### Operating System

Windows

### FastAPI Version

0.75.2

### Python Version

Python 3.10.1

**Comment by jgould22** [comment]
They are essentially the same. Just a little indirection.

To humor you however I did try it and I get a cookie each time.

How are you making your requests ?

**Comment by Naxii-e** [comment]
I guess I was greatly mistaken.
I was using `templates.TemplateResponse()` for the return value.
I changed it to `JSONResponse` and the cookie was set correctly.

Does this mean that `templates.TemplateResponse()` cannot set a cookie?

This is part of the actual codes.

```python

from fastapi import FastAPI, Request, Response

app = FastAPI()
templates = Jinja2Templates(directory='templates/')

@app.get('/login')
async def login(request: Request):
    return templates.TemplateResponse("login.html", context={'request': request})

@app.post('/login')
async def login(request: Request, response: Response, name: str = Form(...), pw: str = Form(...)):
    user = authenticate(name, pw)
    response.set_cookie(key="access", value="aaaaa")
    response.set_cookie(key="refresh", value=str(generate_tokens(user.id, "refresh")))
    return templates.TemplateResponse("login-done.html", context={'request': request, 'name': user.name})
```

**Comment by jgould22** [comment, marked as answer]
You are setting the cookie on the response from params but then creating and returning an entirely new response with your return.
 
Set the cookie on your template response by changing the below
```python
@app.post('/login')
async def login(request: Request, response: Response, name: str = Form(...), pw: str = Form(...)):
    user = authenticate(name, pw)
    response.set_cookie(key="access", value="aaaaa")
    response.set_cookie(key="refresh", value=str(generate_tokens(user.id, "refresh")))
    return templates.TemplateResponse("login-done.html", context={'request': request, 'name': user.name})
```
 to something like 
```python
@app.post('/login')
async def login(request: Request, name: str = Form(...), pw: str = Form(...)):
    user = authenticate(name, pw)
    template_response =  templates.TemplateResponse("login-done.html", context={'request': request, 'name': user.name})
    template_response.set_cookie(key="access", value="aaaaa")
    template_response.set_cookie(key="refresh", value=str(generate_tokens(user.id, "refresh")))
    return template_response
```

**Reply by boludoz** [reply]
You're a fucking genius.

**Comment by Naxii-e** [comment]
It worked perfectly!
I must have done something wrong in the way I was responding.
My apologies.

![image](https://user-images.githubusercontent.com/78034307/166235111-25e74b83-f07f-4c9d-9f0d-accff41132dd.png)


Thank you for your thoughtful response.
